### ISSUE TITLE
Feature Request: List Item Animations

### ISSUE THREAD

**Issue by brianegan** [flutter/flutter#6265]
## Request

When adding, removing, or changing the order of items in a ScrollableList or LazyBlock, the items are simply redrawn in the correct order, but they don't animate in an elegant way.

For example, if a list item is removed, the rest of the items could "slide up" to fill in the gap left behind by the removed item.

If the order is shuffled, the list items could animate to their new position.
## Visual Examples

Video:
https://vimeo.com/183667253

A code example using the Android APIs:
https://github.com/wasabeef/recyclerview-animators
## Nice-to-haves
1. Simply hand a new list of children to a `ScrollableList` and it handles all the animation magic.
2. A set of standard list item animations (Material style animations, iOS style, etc)
3. It would be great to have an extension to define exactly how the animations should be executed. The `RecyclerViewAnimator` class from Android is an example of what I mean. I've used both stock Material transitions and had to hand-write them for the company I work for.
## Current ways to implement

More ideas welcome, but in order to make this affect happen, I was simply going to manually calculate the animations, run them, and THEN update the state of my application through `setState`. An example of this approach can be seen in the Mithril.js documentation:

http://lhorie.github.io/mithril-blog/velocity-animations-in-mithril.html

That said, it would be simpler if this could "just happen" when you pass a new set of data into the `ScrollableList` or `LazyBlock`

**Comment by eseidelGoogle** [comment]
Sounds like some good examples to write for our Gallery.  @HansMuller @Hixie

**Comment by Hixie** [comment]
We could add this to AnimatedList maybe.

**Comment by The-Redhat** [comment]
Hey, anything new on this topic ?

**Comment by sergeymild** [comment]
Yes, it will be very nice to see this feature

**Comment by wemped** [comment]
Attempting to use AnimatedList right now for complex changes in lists and really struggling. Especially because it's missing grouped addition/removals (like appending many items to the list) as well as an `updateItem` method to swap out an item for another, (you can remove then add, but the animation looks odd - especially at the end of a list because the scroll view will scroll up on the remove but not scroll back down for the add). 

I'd love a more robust AnimatedList, features like the previously mentioned `DiffUtil` or handing the list new children and it figures out the animation changes needed would be invaluable to me and my team.

**Comment by CorvetteCole** [comment]
Agreed. Making the jump to Flutter and rebuilding a company native Android app and this is really lacking. I used RecyclerViews extensively and the animations were very very nice

**Comment by mkiisoft** [comment]
Any updates on this?

**Comment by PegasisForever** [comment]
I made a very simple implementation, the list can animate items automatically when data changes.
Code:  [https://gist.github.com/PegasisForever/e4ba8c18068fbfb7fbbabbe610517de1](BetterAnimatedList.dart)
Example: [https://gist.github.com/PegasisForever/fb5b2cfc5194eb0957eb18142de65206](main.dart)
